### ISSUE TITLE
scala3: publishing scala 3 artifacts

### ISSUE THREAD

**Issue by raboof** [akka/akka#30243]
There are 2 ways Akka can support Scala 3: by making sure our 2.13.x artifacts work from Scala 3, and by publishing our own Scala 3 artifacts.

This is an 'umbrella issue' to track progress and issues with the second, more long-term goal: publishing our own Scala 3 artifacts. The other, perhaps more short-term goal is tracked in #30242.

A good first step is compiling modules with Scala 3. You can already compile some modules with `sbt -Dakka.build.scalaVersion=3.0`, and a Travis test makes sure that those remain compilable on Scala 3.

* [x] akka-actor
* [x] Test/akka-actor-tests
* [x] akka-actor-testkit-typed #30395 
* [x] Test/akka-actor-testkit-typed #30423
* [x] akka-actor-typed #30258
* [x] Test/akka-actor-typed-tests  #30423
* [x] akka-bench-jmh #30682 
* [x] akka-cluster/compile #30417 
* [x] akka-cluster/Test/compile #30417 
* [x] akka-cluster/test #30890
* [x] akka-cluster-metrics/Test/compile #30690
* [x] akka-cluster-metrics/test #30891
* [x] akka-cluster-sharding #30712
* [x] akka-cluster-sharding/test #30904
* [x] akka-cluster-sharding-typed #30738
* [x] akka-cluster-sharding-typed/test #30892
* [x] akka-cluster-tools #30683
* [x] akka-cluster-typed #30703 
* [x] akka-coordination #30386
* [x] Test/akka-coordination #30409 
* [x] akka-discovery #30396
* [x] Test/akka-discovery #30410
* [x] akka-distributed-data #30666
* [x] akka-distributed-data/test #30666
* [ ] akka-docs depends on https://github.com/spray/spray-json/pull/342
* [x] akka-persistence #30406
* [ ] akka-persistence-tck #30911
* [x] akka-persistence-testkit #30772
* [x] akka-persistence-typed #30772
* [ ] akka-persistence-typed-tests #30911
* [x] akka-persistence-query #30663
* [x] akka-persistence-shared #30663
* [x] akka-pki #30385
* [x] akka-protobuf #30385
* [x] akka-protobuf-v3 #30385
* [x] akka-remote #30361
* [x] akka-remote/test #30415
* [x] akka-remote-tests/test #30415
* [x] akka-slf4j #30385
* [x] akka-serialization-jackson #30408
* [x] akka-serialization-jackson/Test/compile #30408
* [x] akka-serialization-jackson/test #30670 
* [x] akka-stream #30324
* [x] akka-stream-testkit #30407 
* [ ] akka-stream-typed #30910 
* [x] test:akka-stream-tests-tck #30360
* [x] test:akka-stream-tests #30451
* [x] akka-testkit #30407 
* [x] akka-actor/doc (depends on https://github.com/lampepfl/dotty/issues/13061)

Also:
* [ ] #30245

One challenge is jackson-module-scala will have experimental scala3 support in v2.13.0 (https://github.com/FasterXML/jackson-module-scala#scala-3), but we might not want to update to that just yet (see also e.g. #30343). Since we don't use jackson-module-scala directly, but add it as a transitive dependency for the convenience of downstream projects, as a first step we might just remove it from the dependencies for the Scala 3 build.

**Comment by pjfanning** [comment]
jackson-module-scala 2.13.0-rc1 supports scala 3

**Comment by godenji** [comment]
any ETA on when this will land? A month, 3 months, longer? 

Various libraries in the ecosystem are waiting on a Scala 3 supported Akka release, would be nice to have an idea when this PR will merged.

**Comment by raboof** [comment]
We don't have a projected date. We welcome any help on the remaining tasks above!

**Comment by smarter** [comment]
>  I got them to Sonatype staging but it turns out that doc artifacts are required.

You can cheat and publish an empty doc jar:
```scala
Compile / doc / sources := Seq()
```

> Contributions with solving the Scaladoc issues for Scala 3 would be much welcome.

What issues?

**Comment by erikvanoosten** [comment]
I am trying to build metrics-scala against Akka 2.6.17 with Scala 3. I am getting the following error:

```
[error] error while loading package$,
[error] class file akka/actor/package.class is broken, reading aborted with class dotty.tools.tasty.UnpickleException
[error] TASTy signature has wrong version.
[error]  expected: {majorVersion: 28, minorVersion: 1}
[error]  found   : {majorVersion: 28, minorVersion: 2 [unstable release: 1]}
[error] 
[error] This TASTy file was produced by an unstable release.
[error] To read this TASTy file, your tooling must be at the same version.
[error] The TASTy file was produced by Scala 3.1.1-RC1-bin-20211007-c041327-NIGHTLY-git-c041327.
```

Could Akka be compiled with Scala 3.0.2, or at least 3.1.0?

**Comment by raboof** [comment]
> I am trying to build metrics-scala against Akka 2.6.17 with Scala 3.1.0

Thanks for taking it for a spin!

> I am trying to build metrics-scala against Akka 2.6.17 with Scala 3.1.0
>
>     [error]  expected: {majorVersion: 28, minorVersion: 1}
>     [error]  found   : {majorVersion: 28, minorVersion: 2 [unstable release: 1]}

This might have been fixed with #30817 - could you try a nightly Akka build? https://oss.sonatype.org/content/repositories/snapshots/com/typesafe/akka/ should list them.

> Could Akka be compiled with Scala 3.0.2, or at least 3.1.0?

I'm pretty sure we updated to the -RC version of Scala 3.1.1 because it contains fixes we needed to successfully build Akka with it, though I don't have a complete list of which ones (https://github.com/akka/akka/blob/main/project/Dependencies.scala#L32-L34 mentions https://github.com/lampepfl/dotty/issues/13106 and something around static forwarders, but I suspect there was more). I'm not sure it's worth going back and trying to find workarounds for those issues, probably better to just wait for a new Scala release.

**Comment by erikvanoosten** [comment]
Thanks @raboof, I can confirm that with `2.6.17+25-527571fc-SNAPSHOT` the problem goes away.

**Comment by raboof** [comment]
We have now merged the fix for the last known run-time problem, and the snapshots from `2.6.17+70-9b174a31-SNAPSHOT` onwards should contain it.

That makes it time to close this issue, let's create separate issues for any missing documentation and remaining problems that we might find.

Happy hAkking! :tada:
